FastAPI: Add response wrapper for `SessionRequestContext`

davelopez opened the pull request:
Similar to https://github.com/galaxyproject/galaxy/pull/12531 but in this case for the response object.

This is necessary for migrating `/api/histories/{encoded_history_id}/contents/{encoded_content_id}/display` in https://github.com/galaxyproject/galaxy/pull/12924 because deep down the [display_data](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/data.py#L371) function, and it's children overrides, use the response object through `trans.response` to set or read the headers depending on multiple conditions.

Ideally, we want to completely avoid using `trans.response` in the datatypes module so I will try to explore removing them in a future PR.

## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Add abstract interface for galaxy response

davelopez committed:
Wrap FastAPI/Starlette response

to mimic the behaviour of web/framework/base.py::Response and similar to `GalaxyASGIRequest`

mvdbeek reviewed (commented):
That looks good in principle, but could you add a unit test for this?
It should be possible to use a technique like in https://fastapi.tiangolo.com/tutorial/testing/#using-testclient.

The idea would be to construct a very minimal fastAPI app and then check that the response items are set correctly when you use the app. Let me know if you want me to give this a shot

davelopez committed:
Add unit tests for response wrapper

davelopez committed:
Add get_trans mock and simplify tests

mvdbeek commented on test/unit/webapps/test_response_wrapper.py:
```diff
@@ -0,0 +1,49 @@
+
+@app.get("/test/change_content_type")
+async def change_content_type(trans=Depends(get_trans)):
+    trans.response.set_content_type("test-content-type")
```
Does this work, as temporary workaround ?
```suggestion
async def change_content_type(trans=Depends(get_trans), response: Response):
    trans.response.__response = response
    trans.response.set_content_type("test-content-type")
```
If I understand the linked code correctly the problem is that the response is in a subresponse, which is everything that comes in via dependencies ? So maybe the trick works ?

davelopez replied:
I'm afraid it doesn't work :disappointed:  https://github.com/galaxyproject/galaxy/pull/12929/commits/e6bdf380d4ced93ae6b489b914700b7d87f4dea5
This new response is probably also a subresponse.

davelopez committed:
Try workaround to set response header

davelopez committed:
Fix unit test

To explicitly state the behavior of using `trans.response.set_content_type` when returning the default implicit response
